ols/src/llms/providers/azure_openai.py

```diff
@@ -52,6 +54,55 @@ def default_params(self) -> dict[str, Any]:
             "verbose": False,
         }
 
+        if self.credentials is not None:
+            # if credentials with API key is set, use it to call Azure OpenAI endpoints
+            default_parameters["api_key"] = self.credentials
+        else:
+            # credentials for API key is not set
+            # -> it is needed to retrieve token first
+            access_token = self.retrieve_access_token(azure_config)
+            # set up active directory token to access Azure services, including OpenAI one
+            default_parameters["azure_ad_token"] = access_token.token
+
+        return default_parameters
+
     def load(self) -> LLM:
         """Load LLM."""
         return AzureChatOpenAI(**self.params)
+
+    def retrieve_access_token(
+        self, azure_config: AzureOpenAIConfig
+    ) -> Optional[AccessToken]:
+        """Retrieve access token to call Azure OpenAI."""
+        if azure_config is None:
+            raise ValueError(
+                "Credentials for API token is not set and "
+                "Azure-specific parameters are not provided."
+                "It is not possible to retrieve access token."
+            )
+        if azure_config.tenant_id is None:
+            raise_missing_attribute_error("tenant_id")
+        if azure_config.client_id is None:
+            raise_missing_attribute_error("client_id")
+        if azure_config.client_secret is None:
+            raise_missing_attribute_error("client_secret")
+
+        # everything is there, try to retrieve credential
+        try:
+            credential = ClientSecretCredential(
+                azure_config.tenant_id,
+                azure_config.client_id,
+                azure_config.client_secret,
+            )
+            return credential.get_token("https://cognitiveservices.azure.com/.default")
```

Comment:
@tisnik does this access token ever expire such that OLS might need to retrieve a new one?

Reply:
yes that expire after some time. it is configurable, but definitely there. Fortunately each request basically refresh the token

Reply:
> Fortunately each request basically refresh the token

meaning each request retrieves a new token?  or each request extends the expiration of the token?

Reply:
are we getting a new access token every time because that is part of loading the llm and we load the llm on every OLS request?

Reply:
yes, basically

Reply:
of course in "normal" REST API I'd optimize it probably. but in case of LLM calls that are pretty costly itself (time+CPU time), dunno if it's worth

Reply:
probably something to mark as a TODO followup.  I agree it won't be a bottleneck but it's one more request flow that can go wrong if something is down and we really shouldn't have to be getting a new token on every request.  I assume Azure would prefer we not get a new token every time also, but i guess it also somewhat depends on the expiration period of the token.